src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h

```diff
@@ -0,0 +1,293 @@
+///  |
+/// loop2(A2, I2, D2)
+/// ```
+///
+/// GOTO instructions are tagged as follows:
+///
+/// ```c
+/// foo /* foo */
+///    CALL do_something()                   // loop_id:3 tags:{top-level}
+///    SKIP; // prehead                      // loop_id:3 tags:{top-level}
+/// 1: DECL __not_done1 : signedbv[32]       // loop_id:0 tags:{head}
+///    CALL __not_done1 := not_done1()       // loop_id:0 tags:{body}
+///    IF ¬(__not_done1 ≠ 0) THEN GOTO 7     // loop_id:0 tags:{body,exiting}
+///    CALL do_something1()                  // loop_id:0 tags:{body}
+///    SKIP; // prehead                      // loop_id:0 tags:{body}
+/// 2: DECL __not_done2 : signedbv[32]       // loop_id:1 tags:{head}
+///    CALL __not_done2 := not_done2()       // loop_id:1 tags:{body}
+///    IF ¬(__not_done2 ≠ 0) THEN GOTO 3     // loop_id:1 tags:{body,exiting}
+///    CALL do_something2()                  // loop_id:1 tags:{body}
+///    GOTO 2                                // loop_id:1 tags:{latch}
+/// 3: SKIP                                  // loop_id:0 tags:{body}
+///    DECL __must_break1 : signedbv[32]     // loop_id:0 tags:{body}
+///    CALL __must_break1 := must_break1()   // loop_id:0 tags:{body}
+///    IF ¬(__must_break1 ≠ 0) THEN GOTO 6   // loop_id:0 tags:{body}
+///    SKIP // prehead                       // loop_id:0 tags:{body}
+/// 4: DECL __not_done3 : signedbv[32]       // loop_id:2 tags:{head}
+///    CALL __not_done3 := not_done3()       // loop_id:2 tags:{body}
+///    IF ¬(__not_done3 ≠ 0) THEN GOTO 5     // loop_id:2 tags:{body,exiting}
+///    CALL do_something3()                  // loop_id:2 tags:{body}
+///    GOTO 4                                // loop_id:2 tags:{latch}
+/// 5: SKIP                                  // loop_id:0 tags:{body}
+///    GOTO 7                                // loop_id:0 tags:{body,exiting}
+/// 6: SKIP                                  // loop_id:0 tags:{body}
+///    DEAD __must_break1                    // loop_id:0 tags:{body}
+///    DEAD __not_done2                      // loop_id:0 tags:{body}
+///    GOTO 1                                // loop_id:0 tags:{latch}
+/// 7: SKIP                                  // loop_id:3 tags:{top-level}
+///    DEAD __not_done1                      // loop_id:3 tags:{top-level}
+///    END_FUNCTION                          // loop_id:3 tags:{top-level}
+/// ```
+///
+/// The tags allow to the dynamic frames instrumentation pass to select the
+/// write set instance of the specific loop to instrument the instruction,
+/// and allow the loop contracts instrumentation pass to robustly locate head,
+/// exiting nodes and latch nodes when applying the loop contract
+/// transformation.
+class dfcc_cfg_infot
+{
+public:
+  dfcc_cfg_infot(
+    const irep_idt &function_id,
+    goto_functiont &goto_function,
+    const exprt &top_level_write_set,
+    const dfcc_loop_contract_modet loop_contract_mode,
+    goto_modelt &goto_model,
```

Comment:
You are taking a `function_id`, a `goto_function`, and a `goto_model`. How do these relate? I'd hope that the `goto_function` is the value where `function_id` is the key in `goto_model`'s function map. If so, please omit the `goto_function` parameter and compute `goto_function`. Else you are just asking for inconsistency.

Reply:
Yes, the parameter `goto_function` should be removed. I will fix it.

Reply:
There are places in contract instrumentation where we instrument goto programs snippets that will eventually land in the function named `function_id` but are not there yet. These snippets are temporarily embedded in dummy goto functions because natural loop or may alias functions only accept goto_functions as inputs.
These goto functions are not found in the function table of the goto model. Fresh symbols generated when instrumenting these dummy goto functions have to use `function_id` as base name so it needs to be passed around, so does the goto model because it carries the symbol table which is needed when creating  fresh symbols.

Making the suggested change will force us to rewrite large parts of the instrumentation code in `dfcc_instrumentt` and `dfcc_wrapper_functiont`, or to temporarily insert/remove dummy goto functions in the goto model just to be able to pass the goto model and the key instead of the goto model, the key and the goto function.

Reply:
Actually, @remi-delmas-3000 is correct. And removing `goto_function` will break the tests. I will undo the fix.

Reply:
if you look at `dfcc_instrumentt` in #7541 you'll see the different entry points for instrumentation, hopefully that makes things clearer.

Reply:
I removed `goto_model` and use`symbol_table` instead to make it less confusing.